README.md
# ASCE 7 Response Spectra Comparison Tool

## Overview
This tool is designed for users to compare the ASCE 7-22 response spectra for a given average shear wave velocity, Vs100, that was measured or estimated by correlation. Users can input specific geographic coordinates (latitude and longitude) and select various other inputs and the tool will retrieve the spectral ordinates from the USGS. The tool provides interactive visualizations, a simple map for location verification, and the ability to download the spectra of interest.
## Features
- **Data Scraping**: Automatically scrapes spectral ordinate data from the [USGS Seismic Design Web Services](https://earthquake.usgs.gov/ws/designmaps/) website.
- **User Input**: Allows users to input latitude and longitude and risk category, and other inputs.
- **Interactive Plots**: Visualize the spectral data for easy comparison.
- **Map Integration**: Provides a geographic map to validate the selected location.
- **Maximum Envelope Spectrum**: Generates a spectrum composed of the Vs100, 1.3*Vs100, and Vs100/1.3 as required for estimated shear wave velocity profiles. 
- **Data Download**: Enables users to download the spectral ordinates of the spectra of interest.

## Usage
Access the app with the URL below. Directions for its use are displayed there.

**To use the app**: [Spectra Comparison Tool](link.com).

## Modify the Code

### Prerequisites
- Python 3.11
- Pipenv for dependency management

### Setup
1. Clone the repository:
   ```
   git clone (url)
   ```
2. Navigate to your project directory:
   ```
   cd asce7_spectra_app
   ```
3. Install dependencies using Pipenv:
   ```
   pipenv install
   ```
4. Activate the Pipenv shell:
   ```
   pipenv shell
   ```
5. Start the Streamlit app:
   ```
   streamlit run source/main.py
   ```
6. To quit:
   ```
   ctrl+c
   ```


### Example/ Test run
To confirm that your setup worked, run the app and submit the default settings.

1. Open a new browser window. 
2. At the command line, enter `streamlit run source\main.py`. A browser window will open with the app.
3. Open the sidebar with the little arrow at upper left.
4. Leave the defaults. Select `Check Location` to confirm that the map updates to a point near Issaquah, WA.
5. Select `Submit` to add a plot with 4 spectra, a dataframe, and URLs to the main area.
6. You're now free to change the inputs in the sidebar for your project!


### License
This will be a private repo, right? Need to figure out how to deploy on local server.

## Disclaimer
### From the USGS, the source of the data:
This software is preliminary or provisional and is subject to revision. It is being provided to meet the need for timely best science. The software has not received final approval by the U.S. Geological Survey (USGS). No warranty, expressed or implied, is made by the USGS or the U.S. Government as to the functionality of the software and related material nor shall the fact of release constitute any such warranty. The software is provided on the condition that neither the USGS nor the U.S. Government shall be held liable for any damages resulting from the authorized or unauthorized use of the software.

### From the author:
This tool was developed to aid exploratory analysis of projects. It is subject to revision. Though developed thoughtfully, neither the functionality of the software nor the reliablity of the data are guaranteed. No warranty, expressed or implied, is made by the author. Users are urged to vet the data and information here against a unique source before relying on it for decision making.

Changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)

## Unreleased
- TODO: add error handling for locations outside of 7-22 coverage.
- TODO: add more instructions to use.
- TODO: add testing to ensure accurate webscraping.
- TODO: add some error handling, like for outside of US and there's no plot data.
- TODO: refactor code. Maybe ChatGPT can help?

## 2023 Dec 11

### Added
- one global password to app


## 2023 Nov 11

### Added
- clone the SpectraPlotter repo here for new project.
  
### Changed
- changed pipfile so that python_full_version = "3.11.5" instead of 3.11.6

source/plotting.py
import plotly.graph_objects as go

# Define the plotting function and style
'''
Input:
    fig: an initialized figure, e.g. fig1 = go.Figure()
    df: a dataframe with column 0 called 'Date' and other columns the series of interest.
Output: 
    A plot of the 'measured' and 'estimated' composite spectra

'''
def plot_spectra(fig, all_data, my_df, title):
    # Add the spectrum for the user's 'measured' site class.
    fig.add_trace(
        go.Scatter(
            x=all_data[0]['Periods'],
            y=all_data[0]['Ordinates'],
            name=f"Site Class {all_data[0]['Site Class']} if Vs100 measured"
        )
    )

    # Make the composite spectrum
    # Use the periods from first site class since all the same.
    periods = all_data[0]['Periods'] if all_data else []

    # Get max spectral acceleration at each period.
    max_ords = [max(data['Ordinates'][i] for data in all_data) for i in range(len(periods))]
    
    # Add the composite spectrum to the plot.
    fig.add_trace(
        go.Scatter(
            x=periods,
            y=max_ords,
            name='Composite Spectrum if Vs100 estimated',
            line=dict(dash='dash')
        )
    )

    # Add the composite spectrum to the dataframe.
    my_df['Composite'] = max_ords


    # Set x axis to log and add labels.
    fig.update_xaxes(
            type='log',
            title='Period (s)'
        )
    
    fig.update_yaxes(
            title='Spectral Acceleration (g)'
        )
    
    fig.update_layout(
        title=f'Multiperiod Response Spectra for {title}',
        hovermode='x unified'
    )
 
    return(fig, my_df)

source/functions.py
import scraper

# Construct the URL to be passed to scraper.
def construct_url(latitude, longitude, riskCategory, siteClass, title):
    base_url = "https://earthquake.usgs.gov/ws/designmaps/asce7-22.json?"
    params = {
        'latitude': latitude,
        'longitude': longitude,
        'riskCategory': riskCategory,
        'siteClass': siteClass,
        'title': title
    }
    return base_url + '&'.join(f"{key}={value}" for key, value in params.items())

# Function to remove duplicate site classes from a list of all site classes.
def remove_duplicates(lst):
    seen = set()
    seen_add = seen.add
    return [x for x in lst if not (x in seen or seen_add(x))]

# Function to test construct_url function  
def main():
    # Example inputs - replace with user inputs from Streamlit widgets
    latitude = 34
    longitude = -118
    riskCategory = "III"
    siteClass = "C"
    title = "Example"

    url = construct_url(latitude, longitude, riskCategory, siteClass, title)
    print("the constructed URL is: ", url)
    periods, ordinates = scraper.scrape_data(url)

    if periods and ordinates:
        print("Periods:", periods)
        print("Ordinates:", ordinates)
    else:
        print("Failed to retrieve data.")

if __name__ == "__main__":
    main()

source/make_dataframe.py
import pandas as pd

def make_dataframe(all_data):
    '''
    A function to make a human-friendly table of the scraped site class data.
    Input:
        fig: an array of scraped data containing site class, periods, and Sa.
    Output: 
        A dataframe with 
            header: unique Site Classes in the scraped data set
            col0: periods for spectra
            col1: the 'measured' spectrum ordinates
            col2: the 'composite' spectrum ordinates
    '''
    # Get periods from first site class since same for all.
    periods = all_data[0]['Periods'] if all_data else []
    my_df = pd.DataFrame({'Periods': periods})

    # Add columns for each site class.
    for data in all_data:
        site_class = data['Site Class']
        ordinates = data['Ordinates']
        my_df[site_class] = ordinates
        
    return my_df

source/main.py
# An app to plot the ASCE 7-22 response spectra
# given the user selected location and site class(es).

# Import libraries
import streamlit as st
import hmac
import pandas as pd
import plotly.graph_objects as go
import scraper
import functions
import ASCE
import plotting
import make_dataframe

def check_password():
    """Returns `True` if the user had the correct password."""

    def password_entered():
        """Checks whether a password entered by the user is correct."""
        if hmac.compare_digest(st.session_state["password"], st.secrets["password"]):
            st.session_state["password_correct"] = True
            del st.session_state["password"]  # Don't store the password.
        else:
            st.session_state["password_correct"] = False

    # Return True if the passward is validated.
    if st.session_state.get("password_correct", False):
        return True

    # Show input for password.
    st.text_input(
        "Password", type="password", on_change=password_entered, key="password"
    )
    if "password_correct" in st.session_state:
        st.error("Password incorrect")
    return False

if not check_password():
    st.stop()  # Do not continue if check_password is not True.
# Define a function to show plot, dataframe, and URLs
def do_all_the_processing():
    # Get the value of the slider
    vs100 = st.session_state.vs100_slider
    site_class_user = ASCE.asceTable(vs100)
    st.write(f"The selected Vs100 = {vs100}, Site Class {site_class_user}")

    # calculate 1.3* and /1.3
    vs100_multiplied = round(vs100 * 1.3)
    vs100_divided = round(vs100 // 1.3)

    # fetch site classes for the over/under
    site_class_over = ASCE.asceTable(vs100_multiplied)
    site_class_under = ASCE.asceTable(vs100_divided)
    st.write(f'Vs100 * 1.3 is = {vs100_multiplied}, Site Class {site_class_over}')
    st.write(f'Vs100 / 1.3 is = {vs100_divided}, Site Class {site_class_under}')

    # Make list of site class, site class over, site class under
    all_site_classes = [site_class_user, site_class_over, site_class_under]

    # Get the set of unique site classes.
    site_classes = functions.remove_duplicates(all_site_classes)

    # Construct the URLs
    urls = []
    for site_class in site_classes:
        url = functions.construct_url(
            st.session_state.lat,
            st.session_state.lon,
            st.session_state.risk_category,
            site_class,
            st.session_state.title
        )
        urls.append((url, site_class))

    # When URLs are ready, scrape data
    all_data = []
    for url, site_class in urls:
        periods, ordinates = scraper.scrape_data(url)
        if periods and ordinates:
            all_data.append({
                'Site Class': site_class,
                'Periods': periods,
                'Ordinates': ordinates})

    # Make a human-reader friendly dataframe.
    my_df = make_dataframe.make_dataframe(all_data)
    
    # Plot the user's spectra.
    fig = go.Figure()
    fig, my_df = plotting.plot_spectra(fig, all_data, my_df, st.session_state.title)

    # MAIN AREA: display spectra plot.
    st.plotly_chart(fig)

    # MAIN AREA: display datarame of spectral ordinates.
    st.write('The dataframe contains the spectral ordinates of the plotted data')
    st.dataframe(my_df)

    # MAIN AREA: write out URLs and data source.
    url_list = [url for url, _ in urls]
    url_string = "\n".join(url_list)
    st.write('The data was gathered from the USGS website using these URLs:', url_string)

# Title for app.
st.header('ASCE 7-22 Response Spectra Plotter')

# Check if user has submitted initial input.
if st.session_state.get('user_input'):
    # Display the slider and update it's value in session state.
    vs100_slider = st.slider(
        'Select Vs100 (fps)',
        1, 5000,
        value=st.session_state.get('vs100_slider', 1500), # Use default 1500 if none
        step=1,
        key='vs100_slider'
    )

    if vs100_slider != st.session_state.get('vs100_slider', 1500):
        st.session_state['vs100_slider'] = vs100_slider
        
    # when the slider is changed,
    do_all_the_processing()

# Set a session state.
if 'lat' not in st.session_state:
    st.session_state['lat'] = 47.5678
if 'lon' not in st.session_state:
    st.session_state['lon'] = -122.0123
if 'risk_category' not in st.session_state:
    st.session_state['risk_category'] = 'III'
if 'title' not in st.session_state:
    st.session_state['title'] = 'MyProjectCity'
if 'user_input' not in st.session_state:
    st.session_state['user_input'] = None

# Get initial user input
if st.session_state['user_input'] is None:
    # Add a welcome message and instructions
    st.write('Please submit required data in sidebar to the left.')

# User input location, confirm on map
with st.sidebar:
    st.header('Start by Entering Options Here')

    # SIDEBAR: get latitude and longitude if none, or show defaults.
    st.session_state['lat'] = st.text_input(
        'Enter latitude',
        value=st.session_state['lat']
    )
    st.session_state['lon'] = st.text_input(
        'Enter longitude',
        value=st.session_state['lon']
    )

    # SIDEBAR: submit button for lat, lon
    check_location = st.button('Check Location')

    if check_location:
        try:
            lat_float = float(st.session_state['lat'])
            lon_float = float(st.session_state['lon'])
            loc = pd.DataFrame({'lat': [lat_float], 'lon': [lon_float]})
            st.map(data=loc, zoom=8)
        except ValueError:
            st.error("Please enter valid latitude and longitude values.")

    # SIDEBAR: use a form to get user inputs
    st.header('After confirming location, select spectrum options')
    with st.form('user_inputs'):
        # SIDEBAR: input risk category, only 1.
        risk_categories = ['I', 'II', 'III', 'IV']
        st.session_state['risk_category'] = st.radio(
            'Select risk category',
            risk_categories,
            index=risk_categories.index(st.session_state['risk_category'])
        )
        
        # SIDEBAR: name the location of the project.
        st.session_state['title'] = st.text_input(
            'Project Location',
            value= st.session_state['title']
        )

        # SIDEBAR: submit button
        user_input = st.form_submit_button("SUBMIT")

        # Remember the user inputs that were submitted
        if user_input:
            st.session_state['user_input'] = True

# MAIN AREA: show a disclaimer.
st.write('This tool was developed to aid exploratory analysis of projects.' +
            'It is subject to revision. Though developed thoughtfully, ' +
            'neither the functionality of the software nor the reliablity of ' + 
            'the data are guaranteed. No warranty, expressed or implied, is ' +
            'made by the author. Users are urged to vet the data and ' +
            'information here against a unique source before relying on it ' +
            'for decision making.')